src/compose/ports.ts

```diff
@@ -0,0 +1,326 @@
+	 */
+	public static fromDockerOpts(
+		portBindings: PortBindings,
+	): PortMap[] {
+
+		// A type to help with the parsing
+		interface BindingObject {
+			hostObj: { HostIp: string, HostPort: string }[];
+			internalStr: string;
+			// This is filled in as part of the grouping
+			internalStrMatch?: string[];
+		}
+
+		// Create a list of portBindings, rather than the map (which we can't
+		// order)
+		const portBindingList: BindingObject[] = _.map(portBindings, (hostObj, internalStr) => {
+			return {
+				hostObj,
+				internalStr,
+			};
+		});
+
+		// Now order it by the internal port, so that all ports that are part
+		// of the same range are consecutive
+		const orderedPortBindings = _.sortBy(portBindingList, (binding) => {
+			// Get the port number from the internal string
+			const match = binding.internalStr.match(DOCKER_OPTS_PORTS_REGEX);
+			if (match == null) {
+				throw new Error(`Could not parse docker port output: ${binding.internalStr}`);
+			}
+			// Store the match for later
+			binding.internalStrMatch = match;
+			// return the port to order on
+			return match[1];
+		});
+
+		// At this point we have an ordered list of ports, but they could still
+		// be part of different ranges, based on the host and the protocol, so
+		// we need to further filter them
+		const hostIPBindings: {
+			[hostIP: string]: {
+				[proto: string]: BindingObject[];
+			}
+		} = {};
+
+		_.each(orderedPortBindings, (binding) => {
+			const hostIP = binding.hostObj[0].HostIp;
+			const proto = binding.internalStrMatch![2] || 'tcp';
+
+			if (hostIPBindings[hostIP] == null) {
+				hostIPBindings[hostIP] = {};
+			}
+			if (hostIPBindings[hostIP][proto] == null) {
+				hostIPBindings[hostIP][proto] = [];
+			}
+			hostIPBindings[hostIP][proto].push(binding);
+		});
+
+		// Now separate the ports into ranges
+		let portMaps: PortMap[] = [];
+		// We don't technically need to define this here, but it keeps
+		// TS happy, even though we know it will always be defined by the time
+		// it's used.
+		let currentRange: {
+			internalStart: number | null;
+			internalEnd: number | null;
+			externalStart: number | null;
+			externalEnd: number | null;
+		} = {
+			internalStart: null,
+			internalEnd: null,
+			externalStart: null,
+			externalEnd: null,
+		};
+
+
+		_.each(hostIPBindings, (protoBindings, host) => {
+
+			_.each(protoBindings, (portList, proto) => {
+				currentRange = {
+					internalStart: null,
+					internalEnd: null,
+					externalStart: null,
+					externalEnd: null,
+				};
+
+				portList.forEach((port: BindingObject) => {
+					// We need to check that the internal and external ports
+					// are either one above or one below the last ones, if not
+					// we need to start a new range
+					let internal = parseInt(port.internalStrMatch![1]);
+					let external = parseInt(port.hostObj[0].HostPort);
+
+					// If we are not currently tracking a range
+					if (currentRange.internalStart == null) {
```

Comment:
Maybe you can avoid all this logic, and just turn the `portBindingList` into a `PortMap[]`, and then call `normalisePortMaps` on that? This would make this function way simpler / shorter, *and* it would mean that the same function is used going both ways, which will make chances of comparisons going wrong smaller.

Unless I'm missing some reason why it wouldn't work?